ros2/src/autorccar_costmap/src/costmap.cpp

```diff
@@ -0,0 +1,227 @@
+
+    // Filtering
+    pcl::PassThrough<pcl::PointXYZI> zFilter;
+    zFilter.setInputCloud(cur_point_cloud_transformed);
+    zFilter.setFilterFieldName("z");
+    zFilter.setFilterLimits(0.0, 1.0);
+    zFilter.filter(*cur_point_cloud_filtered);
+
+    for (const auto& pt : *cur_point_cloud_filtered) {
+        // Get cell indices of the pointcloud
+        int x = floor(pt.x / resolution) + global_center_x;
+        int y = floor(pt.y / resolution) + global_center_y;
+
+        // The cell where the lidar point located is occupied.
+        UpdateCell(x, y, P_OCCUPIED);
+
+        // The cells on the bresenham line are free.
+        BresenhamLine(sx, sy, x, y);
+    }
+
+    costmap_flag = false;
+}
+
+void Costmap::BresenhamLine(int sx, int sy, int ex, int ey) {
+    int dx = std::abs(ex - sx);
+    int dy = std::abs(ey - sy);
+
+    // discriminant
+    int d;
+
+    // current point
+    int x = sx;
+    int y = sy;
+
+    std::vector<std::pair<int, int>> line;
+
+    if (dx > dy) {
+        // when 0 < |slope| < 1
+        d = 2 * dy - dx;
+        for (int i = 0; i < dx; i++) {
+            if (d < 0) {
+                d = d + 2 * dy;
+            } else {
+                if (ey > sy) {
+                    y = y + 1;
+                } else {
+                    y = y - 1;
+                }
+                d = d + 2 * (dy - dx);
+            }
+            line.push_back(std::make_pair(x, y));
+
+            if (ex > sx) {
+                x = x + 1;
+            } else {
+                x = x - 1;
+            }
+        }
+    } else {
+        // when |slope| > 1
+        d = 2 * dx - dy;
+        for (int i = 0; i < dy; i++) {
+            if (d < 0) {
+                d = d + 2 * dx;
+            } else {
+                if (ex > sx) {
+                    x = x + 1;
+                } else {
+                    x = x - 1;
+                }
+                d = d + 2 * (dx - dy);
+            }
+            line.push_back(std::make_pair(x, y));
+
+            if (ey > sy) {
+                y = y + 1;
+            } else {
+                y = y - 1;
+            }
+        }
+    }
+
+    // Set costmap.
+    for (unsigned int i = 0; i < line.size(); i++) {
+        UpdateCell(line[i].first, line[i].second, P_FREE);
+    }
+}
+
+bool Costmap::IsInBounds(int x, int y) {
+    if (!(x >= 0 && x < global_size_x)) return false;
+    if (!(y >= 0 && y < global_size_y)) return false;
+    return true;
+}
+
+void Costmap::UpdateCell(int x, int y, double p) {
+    if (IsInBounds(x, y)) {
+        global_costmap(x, y) = global_costmap(x, y) + ProbabilityToLogOdds(p) - ProbabilityToLogOdds(P_UNKNOWN);
+    }
+}
+
+double Costmap::ProbabilityToLogOdds(double p) { return std::log(p / (1.0 - p)); }
+
+double Costmap::LogOddsToProbability(double l) { return std::exp(l) / (1.0 + std::exp(l)); }
+
+struct CostmapInfo Costmap::GetGlobalCostmapInfo() {
+    struct CostmapInfo info;
```

Comment:
```suggestion
CostmapInfo Costmap::GetGlobalCostmapInfo() {
    CostmapInfo info;
```
이거랑 똑같은건가..??
앞에 struct붙이면 뭔가 다른거나 좋은점이 있는건가요 이거?? 정말 몰라서 물어봅니다 :astonished: